content/learn/beginner/inputs-1.md

```diff
@@ -0,0 +1,169 @@
+ structure)
+- `channel` (channel mention or ID, resolves to the channel structure)
+- `role` (role name or ID, resolves as type _\*discordgo.Role_)
+- `duration` (duration that is human-readable, i.e `10h5m` or `10 hour 5 minutes` would both resolve to the same
+   duration)
+
+The `description` is a human-readable description of the argument. This is used in the error message if the argument is
+not valid.
+
+Combining all of this, let's create a custom command that takes two arguments: a coolness level and a user to apply it
+to.
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+```
+
+Note that we are using the `member` type here --- in general, it is recommended to use the `member` type for
+user-related arguments, as it validates that the target user is also a member of the server. There are very limited
+use-cases for `user` and `userid`, which will be left as an exercise to the reader.
+
+### Accessing Arguments
+
+Currently, our code doesn't do anything with the arguments. To access the arguments, we use the `.Get` method on the
+`$args` variable. The syntax is as follows:
+
+```go
+{{ $args.Get <index> }}
+```
+
+The `index` is the index of the argument, starting from 0. The arguments are stored in the order they are defined in the
+`parseArgs` function call. Let us now modify our custom command to access these arguments:
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+
+coolness: {{ $args.Get 0 }}
+member: {{ ($args.Get 1).Nick }}
+```
+
+### Validating Arguments
+
+#### Specifying Valid Ranges
+
+Now, we want to limit the coolness level to a number between 0 and 100. We can do this by adding a simple check:
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+
+{{ if or (gt (toInt $args.Get 0) 100) (lt (toInt $args.Get 0) 0) }}
+Invalid coolness level. Must be between 0 and 100.
+  {{ return }}
+{{ end }}
+
+coolness: {{ $args.Get 0 }}
+member: {{ ($args.Get 1).Nick }}
+```
+
+There is one major things to note about this code: we're starting to repeat a lot of our `$args.Get N` calls! Let's fix
+that first.
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+
+{{ if or (gt (toInt $coolness) 100) (lt (toInt $coolness) 0) }}
+  Invalid coolness level. Must be between 0 and 100.
+  {{ return }}
+{{ end }}
+
+coolness: {{ $args.Get 0 }}
+member: {{ ($args.Get 1).Nick }}
+```
+
+Now, we can make use of another great feature of `parseArgs`: Certain types support additional arguments that can be
+used to validate the input. For example, the `int` type supports two additional arguments that can be used to specify a
+range of valid values. This is especially useful if there are a lot of values to check, as it reduces the amount of code
+we have to write, thus freeing up this valuable space for other things.
```

Comment:
the "especially useful" part is self-evident; I would suggest just saying "can be used to specify a range of valid values, such that the bot will do the validation for us"